common/lockmanager_test.go

```diff
@@ -0,0 +1,362 @@
+
+        threadLockBlock.Done();
+
+        Lock(key1);
+
+        // Signal the child thread locked key1.
+        unlockInChildThreadFirst <- struct{}{};
+
+        err := Unlock(key1);
+        if (err != nil) {
+            test.Fatalf("Failed to unlock for the second time in the child thread.");
+        }
+    }();
+
+    // Wait for the second thread to try to lock key1
+    // while the lock is being used by the first thread.
+    threadLockBlock.Wait();
+    
+    // Small sleep to ensure the thread tries to lock key1.
+    time.Sleep(10 * time.Millisecond);
+
+    // Check if the child thread unlocked key1 before the main thread did.
+    if (len(unlockInChildThreadFirst) > 0) {
+        test.Fatalf("Failed to ensure the main thread unlocked key1 before the child thread unlocked it.");
+    }
+
+    // Unlock for the first time.
+    err := Unlock(key1);
+    if (err != nil) {
+        test.Fatalf("Failed to unlock for the first time in the main thread.");
+    }
+    
+    // Wait for the thread to lock and unlock key1.
+    threadFinishBlock.Wait();
+}
+
+// Lock(1) -> Unlock(1) -> %StaleCheck.
+func TestStaleRetentionWithNonStaleLock(test *testing.T) {
+    clear();
+    defer clear();
+
+    key1 := "testkey1";
+    var threadExecutionBlock sync.WaitGroup;
+
+    // Add key1 to the lockmap and have it unlocked.
+    Lock(key1);
+    Unlock(key1);
+
+    threadExecutionBlock.Add(1);
+
+    go func() {
+        defer threadExecutionBlock.Done();
+
+        RemoveStaleLocksOnce();
+    }();
+
+    // Wait for the thread to finish checking for staleness.
+    threadExecutionBlock.Wait();
+    
+    // Check if the stale lock got removed.
+    _, exists := lockMap.Load(key1);
+    if (!exists) {
+        test.Fatalf("Failed to retain lock even though it was accessed concurrently.");
+    }
+} 
+
+// Lock(1) -> %Stale Check.
+func TestStaleLockRetentionWithLockedKey(test *testing.T) {
+    clear();
+    defer clear();
+
+    key1 := "testkey1";
+    staleDuration := time.Duration(config.STALELOCK_DURATION_SECS.Get()) * time.Second;
+    var staleCheckBlock sync.WaitGroup;
+
+    // Add key1 to the lockmap and have it locked.
+    Lock(key1);
+
+    // Load the lockmap and set the timestamp for the lock to be considered stale.
+    val, _ := lockMap.Load(key1); 
+    val.(*lockData).timestamp = time.Now().Add(-2 * (staleDuration));
+    
+    staleCheckBlock.Add(1);
+
+    // Start a goroutine to check if the lock is stale.
+    go func() {
+        defer staleCheckBlock.Done();
+
+        RemoveStaleLocksOnce();
+    }();
+    
+    // Wait for the thread to finish checking for staleness.
+    staleCheckBlock.Wait();
+
+    // Check if the locked stale lock got removed. 
+    _, exists := lockMap.Load(key1);
+    if (!exists) {
+        test.Fatalf("Failed to retain lock even though it was locked.");
+    }
+}
+
+// Lock(1) -> Unlock(1) -> Make Stale -> %Stale Check first check -> Update timestamp -> %Stale Check second check. 
```

Comment:
In addition to my other comments in the previous test applying,
we do not want to talk about "first" and "second" checks (what happens when we add in a new check between them).
Instead, you can just talk about "parts" or "halves", .e.g., `%StaleCheck(First Part)`.